block/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -308,3 +309,50 @@ getProcessBlocks pm diffusion nodeId desired checkpoints = do
                   else pure False
           when exitedRecovery $
               logInfo "Recovery mode exited gracefully on receiving block we needed"
+
+streamProcessBlocks
+    :: forall ctx m.
+       ( BlockWorkMode ctx m
+       , HasMisbehaviorMetrics ctx
+       )
+    => ProtocolMagic
+    -> Diffusion m
+    -> NodeId
+    -> HeaderHash
+    -> [HeaderHash]
+    -> m ()
+streamProcessBlocks pm diffusion nodeId desired checkpoints = do
+    logInfo "streaming start"
+    r <- Diffusion.streamBlocks diffusion nodeId desired checkpoints (loop (0::Word32) [])
+    case r of
+         Nothing -> do
+             logInfo "streaming not supported, reverting to batch mode"
+             getProcessBlocks pm diffusion nodeId desired checkpoints
+         Just _  -> do
+             logInfo "streaming done"
+             return ()
+  where
+    loop n blocks (streamWindow, wqgM, blockChan) = do
+        streamEntry <- atomically $ readTBQueue blockChan
+        case streamEntry of
+          StreamEnd         -> addBlocks blocks
+          StreamBlock block -> do
+              let batchSize = min 64 streamWindow
+              let !n' = n + 1
```

Comment:
Perhaps it would be best to put the bang on the `n` parameter of `loop`.

Reply:
This one will be forced immediately anyway.

Reply:
Will move `!` to `n`.